test/core/event_engine/posix/event_poller_posix_test.cc

```diff
@@ -513,7 +544,196 @@ TEST_P(EventPollerTest, TestEventPollerHandleChange) {
+      : num_wakeups_(num_wakeups),
+        scheduler_(scheduler),
+        poller_(poller),
+        on_read_(
+            PosixEngineClosure::ToPermanentClosure([this](absl::Status status) {
+              EXPECT_TRUE(status.ok());
+              status = ReadPipe();
+              if (!status.ok()) {
+                // Rarely epoll1 poller may generate an EPOLLHUP - which is a
+                // spurious wakeup. Poll based poller may also likely generate a
+                // lot of spurious wakeups because of the level triggered nature
+                // of poll In such cases do not bother changing the number of
+                // wakeups received.
+                EXPECT_EQ(status, absl::InternalError("Spurious Wakeup"));
+                handle_->NotifyOnRead(on_read_);
+                return;
+              }
+              if (--num_wakeups_ == 0) {
+                // This should invoke the registered NotifyOnRead callbacks with
+                // the shutdown error. When those callbacks call Unref(), the
+                // WakeupFdHandle should call OrphanHandle in the Unref() method
+                // implementation.
+                handle_->ShutdownHandle(absl::InternalError("Shutting down"));
+                Unref();
+              } else {
+                handle_->NotifyOnRead(on_read_);
+                Ref();
+                // Schedule next wakeup to trigger the registered NotifyOnRead
+                // callback.
+                scheduler_->Run(SelfDeletingClosure::Create([this]() {
+                  // Send next wakeup.
+                  EXPECT_TRUE(wakeup_fd_->Wakeup().ok());
+                  Unref();
+                }));
+              }
+            })) {
+    ++kTotalActiveWakeupFdHandles;
+    EXPECT_GT(num_wakeups_, 0);
+    EXPECT_NE(scheduler_, nullptr);
+    EXPECT_NE(poller_, nullptr);
+    wakeup_fd_ = *PipeWakeupFd::CreatePipeWakeupFd();
+    handle_ = poller_->CreateHandle(wakeup_fd_->ReadFd(), "test", false);
+    EXPECT_NE(handle_, nullptr);
+    handle_->NotifyOnRead(on_read_);
+    //  Send a wakeup initially.
+    EXPECT_TRUE(wakeup_fd_->Wakeup().ok());
+  }
+
+  ~WakeupFdHandle() { delete on_read_; }
+
+ private:
+  absl::Status ReadPipe() {
+    char buf[128];
+    ssize_t r;
+    int total_bytes_read = 0;
+    for (;;) {
+      r = read(wakeup_fd_->ReadFd(), buf, sizeof(buf));
+      if (r > 0) {
+        total_bytes_read += r;
+        continue;
+      }
+      if (r == 0) return absl::OkStatus();
+      switch (errno) {
+        case EAGAIN:
+          return total_bytes_read > 0 ? absl::OkStatus()
+                                      : absl::InternalError("Spurious Wakeup");
+        case EINTR:
+          continue;
+        default:
+          return absl::Status(absl::StatusCode::kInternal,
+                              absl::StrCat("read: ", strerror(errno)));
+      }
+    }
+  }
+  void Ref() { ref_count_.fetch_add(1, std::memory_order_relaxed); }
```

Comment:
nit: You can use our internal RefCounted types here, no need to reinvent.

Reply:
Done. I used DualRefCounted now.